### PR TITLE
Forward function update for softplus_function.hpp

### PR DESCRIPTION
**Pull request by nishantkr18**
Hey!
While working on https://github.com/mlpack/mlpack/pull/1912, after training for a few iterations, I noticed that every parameter would turn `null`! Turns out that it was because the `Fn` function for `softplus_function` has a small error.

Since softplus turns close to linear for high input values, we can just return x instead of `log(1 + exp(x))`. Because as x goes higher (around 700), `exp(700)` turns infinity, causing null errors in further calculations.
Further, I find `20` to be an acceptable threshold as done [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/activation.py#L662). 

Let me know your thoughts :)

### CODE REVIEW

**Comment by mrityunjay-tripathi** [comment]
Hey @nishantkr18, I think what you can do for numerical stability is:
```
y = log(1 + e^x)
y = log((1 + e^-x) / e^-x)
y = log(1 + e^-x) + x
```
Now, you can approximate the value for ```e^-x```.
A nice implementation of approximation of ```e^-x``` is given [here](https://github.com/mlpack/mlpack/blob/3ce835cbabd665d40ad11bbb1f841118c7a16702/src/mlpack/methods/ann/layer/log_softmax_impl.hpp#L37-L58).

**Comment by kartikdutt18** [comment]
> Further, I find 20 to be an acceptable threshold as done here.

Hey @nishantkr18, Great work. Do you think we can add a threshold parameter instead which defaults to say 20.

**Comment by nishantkr18** [comment]
> Now, you can approximate the value for `e^-x`.
> A nice implementation of approximation of `e^-x` is given [here](https://github.com/mlpack/mlpack/blob/3ce835cbabd665d40ad11bbb1f841118c7a16702/src/mlpack/methods/ann/layer/log_softmax_impl.hpp#L37-L58).

Hey @mrityunjay-tripathi , on running:
```
  for (double x = 0; x < 100; x += 0.1){
    
    if(std::log(1 + std::exp(-x)) == 0)
      std::cout << "VALUE OF x WHERE IT TURNS 0: " << x << '\n';
  }
```
I find that around `36.8`, the value of the expression comes out to be 0, which means we can safely take the threshold to be at `40`. Besides, `log(1 + exp(x))` gives a more accurate value, and works perfectly in the range of -inf to `700`. 
Also, using `y = log(1 + e^-x) + x` may lead to unnecessary calculations for high values, which ultimately turn out to be `x` :)
Let me know what u think..

**Comment by nishantkr18** [comment]
> > Further, I find 20 to be an acceptable threshold as done here.
> 
> Hey @nishantkr18, Great work. Do you think we can add a threshold parameter instead which defaults to say 20. Let me know what you think.

Yup, could be done..

**Comment by kartikdutt18** [comment]
> Yup, could be done..

Great, somethings to keep in mind then :

1. It should be the last parameter for backward compatibility.
2. Serialize the new parameter.
3. Add a test to show a point where current implementation would have given a NULL but yours doesn't.

I think that should cover it all. Probably you know all this but I guess it doesn't hurt to point them out.

**Comment by mrityunjay-tripathi** [comment]
@nishantkr18: Definitely. I thought you need that for large values (like 700). As `log(1 + e^x)` for large values would turn to `nan` but `log(1 + e^-x) +x` will not. But later realized you are using a threshold value. That's fine if you put the threshold.

**Comment by nishantkr18** [comment]
> > Yup, could be done..
> 
> Great, somethings to keep in mind then :
> 
> 1. It should be the last parameter for backward compatibility.
> 2. Serialize the new parameter.
> 3. Add a test to show a point where current implementation would have given a NULL but yours doesn't.
> 
> I think that should cover it all. Probably you know all this but I guess it doesn't hurt to point them out.

@kartikdutt18 , I was thinking, whats the point of adding a parameter? After all, above `36.8`, any value of threshold will result in the same value (an exact linear function), and below it, the implementation will no longer be accurate or correct for that matter.. Thats what I think, though I might be wrong.. What are your thoughts?

**Comment by kartikdutt18** [comment]
> I was thinking, whats the point of adding a parameter? 

The reasons for adding a parameter are mentioned below:
1. I wouldn't necessarily prefer hardcoded values in codebase (at least as far as I know, aside from universal and experimental  constants we don't use hardcoded values anywhere. And if we do use them we generally make experimental constants accessible via setters and getters ).
2. Many functions use approximation / Thresholding and each of them have a parameter.
3. Pytorch also provides access to this parameter.

I am now copying some portion from [why we don't hardcode values](https://softwareengineering.stackexchange.com/questions/67982/is-it-ever-a-good-idea-to-hardcode-values-into-our-applications):

> I cringe whenever I see statements like command.Timeout = 600. Why 600? Who decided that? Was it timing out before, and someone raised the timeout as a hack instead of fixing the underlying performance issue? Or is it actually some known and documented expectation for processing time?

> If the meaning isn't instantly recognizable. It's safe to assume that everybody knows what 3.14159265... is. It isn't safe to assume that everybody will recognize the gravitational constant, or even pi/2. ("Everybody" here does depend on the nature of the software. Systems programmers can be expected to know the octal representation of Unix permission bits or the like. In naval/marine architecture software, checking the Froude number of a proposed hull and speed to see if it's 1.1 or greater might be perfectly self-explanatory to anybody who should be working on it.)
> If the context isn't recognizable. Everybody knows there are 60 minutes in an hour, but multiplying or dividing by 60 may be unclear if there's no immediate indications that the quantity is a time value or a rate value.

At least to  increase readability and for the sake of consistency we need to add a parameter. As far as the value = 20 is concerned, I would run some tests to figure out why 20 and why not say 25 ?
I might do this tonight, though no promises :)

**Comment by nishantkr18** [comment]
@kartikdutt18 , Sorry for the late reply.
I have made the changes, althought I am not sure how to serialize the parameter. I am hoping the current changes should suffice..
Also, I agree with you on making tests, but there are no tests for any of the activations, as far as I can tell. So, it would make sense to add tests for all activations then, what do u think? Should it be considered for a separate PR?

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp [hunk 1]
```diff
@@ -47,28 +47,32 @@ class SoftplusFunction
...
    * Computes the softplus function.
    *
    * @param x Input data.
    * @param y The resulting output activation.
+   * @param threshold The threshold value above which softplus behaves linear.
    */
   template<typename InputType, typename OutputType>
-  static void Fn(const InputType& x, OutputType& y)
+  static void Fn(const InputType& x, OutputType& y,
+                  const double threshold = 40.0)
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/softplus_function.hpp:70]: Hey, Could you fix the spacing here. Thanks a lot.

**Comment by nishantkr18** [comment]
> Also you can refer to hard shrink layer as well as base layer for better understanding.

@kartikdutt18  From what I could understand, hard shrink does not have static functions and has constructors for initializing parameters. But softplus uses static functions. So are you suggesting to add softplus to `layers`, instead of `activation_functions`, then do the required changes?

**Comment by kartikdutt18** [comment]
> constructors for initializing parameters. But softplus uses static functions. So are you suggesting to add softplus to layers, instead of activation_functions, then do the required change?

Yes, Sorry I forgot to reply.

**Comment by rcurtin** [comment]
(closed and reopened to re-trigger the Azure pipelines build, which hopefully won't fail this time)

**Comment by kartikdutt18** [comment]
Hey @nishantkr18, sorry I got kind of caught up in my exams. They are finally over. Do you need any help with this or any clarification. Kindly let me know.

Thanks a lot.
Regards,
Kartik.

**Comment by nishantkr18** [comment]
> sorry I got kind of caught up in my exams. They are finally over.

Hope they went well :smile:

> Do you need any help with this or any clarification. Kindly let me know.

Sorry for the delay in this PR. I'll try to get this done asap.
I see that SoftplusFunction's individual `Fn` and `Deriv` are used in `ann/rbm/spike_slab_rbm_impl.hpp` and `reparametrization_impl.hpp`. But, if we convert softplus_function to a class having non-static functions, we'd need to create objects to access those functions I think (unless there is some other way).
In that case, I think we can choose from two options:
1. Use `y = log(1 + e^-x) + x` as , which wont require any threshold, as suggested by @mrityunjay-tripathi 
2. We leave it upto the user to use `SoftplusFunction::Fn` and `SoftplusFunction::Deriv` separately instead of adding SoftplusFunction as a layer, if they want to go with a custom threshold.

**Comment by kartikdutt18** [comment]
> I see that SoftplusFunction's individual Fn and Deriv are used in ann/rbm/spike_slab_rbm_impl.hpp and reparametrization_impl.hpp.

I took a look at that and they can be changes `Forward` and `Backward` very easily but yes we would need objects.

> We leave it upto the user to use SoftplusFunction::Fn and SoftplusFunction::Deriv separately instead of adding SoftplusFunction as a layer, if they want to go with a custom threshold.

This would end up adding redundant code to perform the same thing. Because we have to maintain each line of code, redundancy in terms of lines of code should be avoided.

Another solution and I am not sure about this would be :

```
if (x < DBL_MAX)
       return x > -DBL_MAX ? (isnan(std::log(1 + std::exp(x)) ? x :  std::log(1 + std::exp(x)))) : 0;
     return 1.0;
```

Change this in forward, let me know if it works. (Again, I'm not sure about this one).

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp [hunk 2]
```diff
@@ -47,13 +47,14 @@ class SoftplusFunction
...
-  static double Fn(const double x)
+  static double Fn(const double x, const double threshold = 40.0)
   {
-    if (x > 20.0)
+    if (x > threshold)
       return x;
-    else 
+    else
       return std::log(1 + std::exp(x));
   }
```
Comment (mrityunjay-tripathi) [inline src/mlpack/methods/ann/activation_functions/softplus_function.hpp]: Personally, I don't think using threshold as an argument in `Fn` is a great way. What might have been a way to do this is to define a variable in `private` and implement getter and setter methods (with legit default threshold value). And if no threshold value is to be used at all then the implementation could be -

```
static double Fn(const double x)
{
  if (x > 0)
    return std::log(1 + std::exp(-x)) + x;
  else
    return std::log(1 + std::exp(x));
} 
```
You can also use fast approximation of e^-x, referenced in one of my previous comments. Experimenting might reveal if approximation is better choice or not? Note that this is just to evade unnecessary clutter of using `threshold`.

**Comment by kartikdutt18** [comment]
> Personally, I don't think using threshold as an argument in Fn is a great way. What might have been a way to do this is to define a variable in private and implement getter and setter methods (with legit default threshold value).

We can't add it anyway. Getter and setter will only work if it's a layer. This is alias for base layer. You can refer to it and understand it why the above approach won't work. I think I left a comment earlier about this.

**Comment by nishantkr18** [comment]
> if (x < DBL_MAX)
>        return x > -DBL_MAX ? (isnan(std::log(1 + std::exp(x)) ? x :  std::log(1 + std::exp(x)))) : 0;
>      return 1.0;
> ```

I too thought of using this on the first place, but thought to go with pytorch's way. Anyways, I'll try and let u know

**Comment by mrityunjay-tripathi** [comment]
>We can't add it anyway. Getter and setter will only work if it's a layer.

Yes, I understand that. I said it could have been a method but unfortunately not.

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp [hunk 3]
```diff
@@ -47,32 +47,30 @@ class SoftplusFunction
...
+  static double Fn(const double x)
   {
-    if (x > threshold)
-      return x;
-    else
-      return std::log(1 + std::exp(x));
+    const double val = std::log(1 + std::exp(x));
+    if (x < DBL_MAX)
+      return x > -DBL_MAX ? (std::isnan(val) || std::isinf(val) ? x : val)
+                          : 0;
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/softplus_function.hpp:57]: Do you think if conditions would look better here?

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp [hunk 4]
```diff
@@ -51,8 +51,10 @@ class SoftplusFunction
    */
   static double Fn(const double x)
   {
+    const double val = std::log(1 + std::exp(x));
     if (x < DBL_MAX)
-      return x > -DBL_MAX ? std::log(1 + std::exp(x)) : 0;
+      return x > -DBL_MAX ? (std::isnan(val) || std::isinf(val) ? x : val)
+          : 0;
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/activation_functions/softplus_function.hpp:57]: I'd much prefer simple `if-else` loops in place of nested conditionals. Improves readability.

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp [hunk 5]
```diff
@@ -102,7 +104,9 @@ class SoftplusFunction
    */
   static double Inv(const double y)
   {
-    return y > 0 ? arma::trunc_log(arma::trunc_exp(y) - 1) : 0;
+    const double val = std::log(std::exp(y) - 1);
+    return y > 0 ? (std::isnan(val) || std::isinf(val) ? y : val)
+        : 0;
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/activation_functions/softplus_function.hpp:109]: Same here.

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp [hunk 6]
```diff
@@ -51,8 +51,10 @@ class SoftplusFunction
    */
   static double Fn(const double x)
   {
+    const double val = std::log(1 + std::exp(x));
     if (x < DBL_MAX)
-      return x > -DBL_MAX ? std::log(1 + std::exp(x)) : 0;
+      return x > -DBL_MAX ? (std::isnan(val) || std::isinf(val) ? x : val)
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/activation_functions/softplus_function.hpp:56]: Please make us of `std::isfinite` in place of two checks. Refer [here](https://en.cppreference.com/w/cpp/numeric/math/isfinite)

Comment (nishantkr18) [reply src/mlpack/methods/ann/activation_functions/softplus_function.hpp]: Thanks for that!

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp [hunk 7]
```diff
@@ -51,9 +51,21 @@ class SoftplusFunction
    */
   static double Fn(const double x)
   {
+    const double val = std::log(1 + std::exp(x));
     if (x < DBL_MAX)
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/activation_functions/softplus_function.hpp:55]: Why this check? If you're checking the value of the outputs, this becomes redundant.

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp [hunk 8]
```diff
@@ -102,7 +114,16 @@ class SoftplusFunction
    */
   static double Inv(const double y)
   {
-    return y > 0 ? arma::trunc_log(arma::trunc_exp(y) - 1) : 0;
+    const double val = std::log(std::exp(y) - 1);
+    if (y > 0)
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/activation_functions/softplus_function.hpp:118]: This check is redundant, as the `Softplus` function will never return negative values. So the only other case is zero, which is already handled below.

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp [hunk 9]
```diff
@@ -51,9 +51,21 @@ class SoftplusFunction
...
+        if (std::isfinite(val))
+          return val;
+        else
+          return x;
+      }
+      else
+        return 0;
+    }
+    else
+      return 1.0;
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/activation_functions/softplus_function.hpp:68]: Why should the value of `log(1 + exp(x))` be `1` when the input `x` is a big integer?

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp [hunk 10]
```diff
@@ -51,9 +51,21 @@ class SoftplusFunction
...
     if (x < DBL_MAX)
-      return x > -DBL_MAX ? std::log(1 + std::exp(x)) : 0;
-    return 1.0;
+    {
+      if (x > -DBL_MAX)
+      {
+        if (std::isfinite(val))
+          return val;
+        else
+          return x;
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/activation_functions/softplus_function.hpp:62]: Please make use of additional braces `{}` with if-else statements.

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp (hunk 8 above)
Comment (nishantkr18) [reply src/mlpack/methods/ann/activation_functions/softplus_function.hpp:118]: Yeah, Ur right. Since it was present from before, I thought not to change it. Now, I guess I should..

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp (hunk 9 above)
Comment (nishantkr18) [reply src/mlpack/methods/ann/activation_functions/softplus_function.hpp:68]: I too was wondering about that. Thought maybe it's made in use somewhere in the code, Ill change and see if any test fails..

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp (hunk 10 above)
Comment (nishantkr18) [reply src/mlpack/methods/ann/activation_functions/softplus_function.hpp:62]: Actually I'm note sure.. Its written `To conserve lines, refrain from using braces if the if, while, or for statement applies to a single statement.` [here](https://github.com/mlpack/mlpack/wiki/DesignGuidelines#brace-placement).. Should I add braces?

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp (hunk 7 above)
Comment (nishantkr18) [reply src/mlpack/methods/ann/activation_functions/softplus_function.hpp:55]: Let me change it and see if any test fails..

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp (hunk 9 above)
Comment (ShikharJ) [reply src/mlpack/methods/ann/activation_functions/softplus_function.hpp:68]: @zoq Since you were involved with this code, maybe you could provide a rationale for this?

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp (hunk 10 above)
Comment (ShikharJ) [reply src/mlpack/methods/ann/activation_functions/softplus_function.hpp:62]: If you place braces right next to the `if-else` endings, you'll only require one extra line. And that's fine, if it improves readability. Compilers are indifferent to the number of lines used.

File: src/mlpack/methods/ann/activation_functions/softplus_function.hpp (hunk 9 above)
Comment (zoq) [reply src/mlpack/methods/ann/activation_functions/softplus_function.hpp]: There is no rational explanation, it's not correct.

**Comment by ShikharJ** [comment]
Thanks @nishantkr18 for the contribution!
